Add torch._C.ScriptDict

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52659 [JIT] Add torch._C.ScriptDict**

**Summary**
This commit adds `torch._C.ScriptDict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch._C.ScriptDict` in TorchScript are visible in
Python even when it is not returned from the function. Instances can be
constructed by passing an instance of a Python dictionary to
`torch.jit.script`. In the case of an empty dictionary, its type is
assumed to be `Dict[str, Tensor]` to be consistent with the handling of
empty dictionaries in TorchScript source code.

`torch.jit.dict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch._C.ScriptDict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

Differential Revision: [D27211605](https://our.internmc.facebook.com/intern/diff/D27211605)

Commit:
[WIP][JIT] Add `torch.jit.dict`

**Summary**
This commit adds `torch.jit.dict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch.jit.dict` in TorchScript are visible in
Python even when it is not returned from the function.

`torch.jit.dict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch.jit.dict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

[ghstack-poisoned]

SplitInfinity commented:
@Lilyjjo 

It's not like *actually* ready for review in the sense that I need to polish some things up, add comments, etc. but I just wanted to bring this work to your attention.

Commit:
Update on "[WIP][JIT] Add `torch.jit.dict`"


**Summary**
This commit adds `torch.jit.dict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch.jit.dict` in TorchScript are visible in
Python even when it is not returned from the function.

`torch.jit.dict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch.jit.dict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Add `torch.jit.dict`"


**Summary**
This commit adds `torch.jit.dict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch.jit.dict` in TorchScript are visible in
Python even when it is not returned from the function.

`torch.jit.dict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch.jit.dict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Add `torch.jit.dict`"


**Summary**
This commit adds `torch.jit.dict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch.jit.dict` in TorchScript are visible in
Python even when it is not returned from the function.

`torch.jit.dict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch.jit.dict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add `torch.jit.dict`"


**Summary**
This commit adds `torch.jit.dict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch.jit.dict` in TorchScript are visible in
Python even when it is not returned from the function.

`torch.jit.dict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch.jit.dict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add `torch.jit.dict`"


**Summary**
This commit adds `torch.jit.dict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch.jit.dict` in TorchScript are visible in
Python even when it is not returned from the function.

`torch.jit.dict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch.jit.dict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add `torch.jit.dict`"


**Summary**
This commit adds `torch.jit.dict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch.jit.dict` in TorchScript are visible in
Python even when it is not returned from the function.

`torch.jit.dict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch.jit.dict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add `torch.jit.dict`"


**Summary**
This commit adds `torch.jit.dict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch.jit.dict` in TorchScript are visible in
Python even when it is not returned from the function.

`torch.jit.dict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch.jit.dict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add `torch.jit.dict`"


**Summary**
This commit adds `torch.jit.dict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch.jit.dict` in TorchScript are visible in
Python even when it is not returned from the function.

`torch.jit.dict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch.jit.dict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add torch._C.ScriptDict"


**Summary**
This commit adds `torch._C.ScriptDict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch._C.ScriptDict` in TorchScript are visible in
Python even when it is not returned from the function. Instances can be
constructed by passing an instance of a Python dictionary to
`torch.jit.script`. In the case of an empty dictionary, its type is
assumed to be `Dict[str, Tensor]` to be consistent with the handling of
empty dictionaries in TorchScript source code.

`torch.jit.dict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch._C.ScriptDict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

The performance benefits have also been tested vs. trunk on a simple
benchmark that creates a large dict and adds an element to it in
TorchScript:
```
@torch.jit.script
def dict_append(a: Dict[int, int], k: int, v: int):
    a[k] = v
    return a

    d = {i: i for i in range(1000000)}
    t1 = time.monotonic()
    d = dict_append(d, -1, -1)
    t2 = time.monotonic()
    print(t2-t1)
```
This involves two copies (Python -> TS -> Python) on trunk and only one
with this commit. This commit speeds up this sample by 75%.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add torch._C.ScriptDict"


**Summary**
This commit adds `torch._C.ScriptDict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch._C.ScriptDict` in TorchScript are visible in
Python even when it is not returned from the function. Instances can be
constructed by passing an instance of a Python dictionary to
`torch.jit.script`. In the case of an empty dictionary, its type is
assumed to be `Dict[str, Tensor]` to be consistent with the handling of
empty dictionaries in TorchScript source code.

`torch.jit.dict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch._C.ScriptDict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

The performance benefits have also been tested vs. trunk on a simple
benchmark that creates a large dict and adds an element to it in
TorchScript:
```
@torch.jit.script
def dict_append(a: Dict[int, int], k: int, v: int):
    a[k] = v
    return a

    d = {i: i for i in range(1000000)}
    t1 = time.monotonic()
    d = dict_append(d, -1, -1)
    t2 = time.monotonic()
    print(t2-t1)
```
This involves two copies (Python -> TS -> Python) on trunk and only one
with this commit. This commit speeds up this sample by 75%.

[ghstack-poisoned]

SplitInfinity commented on torch/csrc/jit/python/pybind_utils.cpp:
```diff
@@ -136,6 +136,17 @@ IValue toIValue(py::handle obj, const TypePtr& type, c10::optional<int32_t> N) {
     }
     case TypeKind::DictType: {
       const auto& dict_type = type->expect<DictType>();
+
+      // If the object is a ScriptDict, retrieve the c10::Dict
+      // instance inside it.
+      try {
+        auto script_dict = py::cast<ScriptDict>(obj);
+        return script_dict.dict_;
+      } catch (py::cast_error& e) {
+      }
+
+      // If not (i.e. it is a regular Python dictionary), make a new
+      // c10::Dict.
```
Add a warning telling users to script their dictionaries instead.

Commit:
Update on "[JIT] Add torch._C.ScriptDict"


**Summary**
This commit adds `torch._C.ScriptDict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch._C.ScriptDict` in TorchScript are visible in
Python even when it is not returned from the function. Instances can be
constructed by passing an instance of a Python dictionary to
`torch.jit.script`. In the case of an empty dictionary, its type is
assumed to be `Dict[str, Tensor]` to be consistent with the handling of
empty dictionaries in TorchScript source code.

`torch.jit.dict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch._C.ScriptDict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

The performance benefits have also been tested vs. trunk on a simple
benchmark that creates a large dict and adds an element to it in
TorchScript:
```
@torch.jit.script
def dict_append(a: Dict[int, int], k: int, v: int):
    a[k] = v
    return a

    d = {i: i for i in range(1000000)}
    t1 = time.monotonic()
    d = dict_append(d, -1, -1)
    t2 = time.monotonic()
    print(t2-t1)
```
This involves two copies (Python -> TS -> Python) on trunk and only one
with this commit. This commit speeds up this sample by 75%.

[ghstack-poisoned]

SplitInfinity commented on torch/csrc/jit/python/python_dict.cpp:
```diff
@@ -0,0 +1,198 @@
+            } catch (const std::out_of_range& e) { // Key doesn't exist.
+              throw py::key_error();
+            } catch (const py::cast_error& e) { // Key isn't the right type.
+              throw py::key_error();
```
I just found out `TypeError` exists and that's what is thrown when you try to index a list with something that's not an integer. Perhaps that would be more appropriate here?

```suggestion
              throw py::type_error();
```

gmagogsfm replied:
Yeah, `TypeError` seems more appropriate here. I think your test verifies that same exception type is thrown between eager and scripted dict, maybe we should have a test to cover/verify that?

Also, a minor suggestion, this try-catch-catch struct should be broken down into two try-catches. First one for key object to IValue and `TypeError` should be caught there, second one covers key lookup.

SplitInfinity replied:
> Also, a minor suggestion, this try-catch-catch struct should be broken down into two try-catches. First one for key object to IValue and TypeError should be caught there, second one covers key lookup.

Why is that? Doesn't `try-catch` go through all the catch blocks in order until one of them "works"? Similar to `switch`?

gmagogsfm replied:
Oh, sorry for not being clear, it is a really minor concern. 

It seems that you want TypeError to catch exceptions rising from converting Python object for key to IValue, and you want `out_of_range`/`KeyError` to catch non-existent keys in dictionary. But, let's say there is a `out_of_range` rising from casting Python object key to IValue, then the exception would become a `KeyError` while it should really be `TypeError`.

SplitInfinity replied:
Yeah that's true

SplitInfinity commented on torch/csrc/jit/python/python_dict.cpp:
```diff
@@ -0,0 +1,198 @@
+              return toPyObject(self->contains(
+                  toIValue(std::move(key), self->type()->getKeyType())));
+            } catch (const py::cast_error& e) {
+              throw py::key_error();
```
```suggestion
              throw py::type_error();
```

SplitInfinity commented on torch/csrc/jit/python/python_dict.cpp:
```diff
@@ -0,0 +1,198 @@
+            try {
+              key_ivalue = toIValue(std::move(key), self->type()->getKeyType());
+            } catch (const py::cast_error& e) {
+              throw py::key_error();
```
```suggestion
              throw py::type_error();
```

gmagogsfm replied:
ditto

SplitInfinity commented on torch/csrc/jit/python/python_dict.cpp:
```diff
@@ -0,0 +1,198 @@
+              value_ivalue =
+                  toIValue(std::move(value), self->type()->getValueType());
+            } catch (const py::cast_error& e) {
+              throw py::value_error();
```
```suggestion
              throw py::type_error();
```

SplitInfinity commented on torch/csrc/jit/python/python_dict.cpp:
```diff
@@ -0,0 +1,198 @@
+            try {
+              key_ivalue = toIValue(std::move(key), self->type()->getKeyType());
+            } catch (const py::cast_error& e) {
+              throw py::key_error();
```
```suggestion
              throw py::type_error();
```

gmagogsfm commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -696,13 +697,7 @@ inline py::object toPyObject(IValue ivalue) {
-          toPyObject(IValue{pair.value()});
-    }
-    return std::move(py_dict);
+    return py::cast(ScriptDict(ivalue));
```
Curious if this has caused any issues in internal CI.

SplitInfinity replied:
It has. I didn't check them all, but I think they should be addressable.

gmagogsfm commented on torch/csrc/jit/python/python_dict.cpp:
```diff
@@ -0,0 +1,198 @@
+        } else {
+          // If is empty, assume the type is Dict[str, Tensor] as is done in
+          // TorchScript code.
+          type = DictType::create(StringType::get(), TensorType::getInferred());
```
@ansley Note that this is another place you might want to pay attention to when inferring type for `{}` and `[]`

gmagogsfm commented on torch/csrc/jit/python/python_dict.cpp:
```diff
@@ -0,0 +1,198 @@
+              throw py::key_error();
+            }
+          },
+          py::return_value_policy::
```
TIL

gmagogsfm commented on torch/jit/_script.py:
```diff
@@ -783,11 +784,36 @@ def call_prepare_scriptable_func_impl(obj, memo):
+        and can be passed between Python and TorchScript with reference semantics and
+        zero copy overhead.
+    """
+    # TODO: Does it make sense to pass in loc=None here?
```
I think so. The only meaningful location to give is "current line", which is naturally available when error is thrown I think.

gmagogsfm reviewed (approved):
LGTM!

Commit:
Update on "[JIT] Add torch._C.ScriptDict"


**Summary**
This commit adds `torch._C.ScriptDict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch._C.ScriptDict` in TorchScript are visible in
Python even when it is not returned from the function. Instances can be
constructed by passing an instance of a Python dictionary to
`torch.jit.script`. In the case of an empty dictionary, its type is
assumed to be `Dict[str, Tensor]` to be consistent with the handling of
empty dictionaries in TorchScript source code.

`torch._C.ScriptDict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch._C.ScriptDict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

The performance benefits have also been tested vs. trunk on a simple
benchmark that creates a large dict and adds an element to it in
TorchScript:
```
@torch.jit.script
def dict_append(a: Dict[int, int], k: int, v: int):
    a[k] = v
    return a

    d = {i: i for i in range(1000000)}
    t1 = time.monotonic()
    d = dict_append(d, -1, -1)
    t2 = time.monotonic()
    print(t2-t1)
```
This involves two copies (Python -> TS -> Python) on trunk and only one
with this commit. This commit speeds up this sample by 75%.

Differential Revision: [D27211605](https://our.internmc.facebook.com/intern/diff/D27211605)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add torch._C.ScriptDict"


**Summary**
This commit adds `torch._C.ScriptDict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch._C.ScriptDict` in TorchScript are visible in
Python even when it is not returned from the function. Instances can be
constructed by passing an instance of a Python dictionary to
`torch.jit.script`. In the case of an empty dictionary, its type is
assumed to be `Dict[str, Tensor]` to be consistent with the handling of
empty dictionaries in TorchScript source code.

`torch._C.ScriptDict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch._C.ScriptDict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

The performance benefits have also been tested vs. trunk on a simple
benchmark that creates a large dict and adds an element to it in
TorchScript:
```
@torch.jit.script
def dict_append(a: Dict[int, int], k: int, v: int):
    a[k] = v
    return a

    d = {i: i for i in range(1000000)}
    t1 = time.monotonic()
    d = dict_append(d, -1, -1)
    t2 = time.monotonic()
    print(t2-t1)
```
This involves two copies (Python -> TS -> Python) on trunk and only one
with this commit. This commit speeds up this sample by 75%.

Differential Revision: [D27211605](https://our.internmc.facebook.com/intern/diff/D27211605)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add torch._C.ScriptDict"


**Summary**
This commit adds `torch._C.ScriptDict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch._C.ScriptDict` in TorchScript are visible in
Python even when it is not returned from the function. Instances can be
constructed by passing an instance of a Python dictionary to
`torch.jit.script`. In the case of an empty dictionary, its type is
assumed to be `Dict[str, Tensor]` to be consistent with the handling of
empty dictionaries in TorchScript source code.

`torch._C.ScriptDict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch._C.ScriptDict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

The performance benefits have also been tested vs. trunk on a simple
benchmark that creates a large dict and adds an element to it in
TorchScript:
```
@torch.jit.script
def dict_append(a: Dict[int, int], k: int, v: int):
    a[k] = v
    return a

    d = {i: i for i in range(1000000)}
    t1 = time.monotonic()
    d = dict_append(d, -1, -1)
    t2 = time.monotonic()
    print(t2-t1)
```
This involves two copies (Python -> TS -> Python) on trunk and only one
with this commit. This commit speeds up this sample by 75%.

Differential Revision: [D27211605](https://our.internmc.facebook.com/intern/diff/D27211605)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add torch._C.ScriptDict"


**Summary**
This commit adds `torch._C.ScriptDict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch._C.ScriptDict` in TorchScript are visible in
Python even when it is not returned from the function. Instances can be
constructed by passing an instance of a Python dictionary to
`torch.jit.script`. In the case of an empty dictionary, its type is
assumed to be `Dict[str, Tensor]` to be consistent with the handling of
empty dictionaries in TorchScript source code.

`torch._C.ScriptDict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch._C.ScriptDict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

Differential Revision: [D27211605](https://our.internmc.facebook.com/intern/diff/D27211605)

[ghstack-poisoned]

SplitInfinity commented:
@gmagogsfm 

Re-requested review as discussed offline. The only thing that has changed is that `toPyObject` always returns a `py::dict`. This preserves reference semantics in the simplest case (i.e. user passes in a `ScriptDict` to a scripted function and does not return it) but breaks nested dictionary reference semantics. To reiterate, this simplification is to allow the foundational changes (e.g. `ScriptDict` class) to land without having to debug and fix every single fbcode failure.

SplitInfinity commented on torch/csrc/jit/python/python_dict.h:
```diff
@@ -0,0 +1,128 @@
+  }
+
+  // A c10::Dict instance that holds the actual data.
+  IValue dict_;
```
I tried this but was not able to default construct `GenericDict`. Let me try again.

SplitInfinity replied:
Okay so it worked, but with the downside that I had to initialize the `dict_` as `dict_(AnyType::get(), AnyType::get())` because there is no default constructor for `GenericDict`. Let me know what you think!

SplitInfinity commented on torch/csrc/jit/python/python_dict.cpp:
```diff
@@ -0,0 +1,205 @@
+      .def(
+          "__repr__",
+          [](const std::shared_ptr<ScriptDict>& self) {
+            return toPyObject(self->repr());
```
Hmm, I'm not sure if they are, I did it mostly out of habit. If pybind tries to convert them, that will be a no-op, right?

gmagogsfm replied:
Should be

SplitInfinity commented on torch/jit/_script.py:
```diff
@@ -885,7 +888,30 @@ def _script_pdt(obj, optimize=None, _frames_up=0, _rcb=None, example_inputs: Opt
+    return torch._C.ScriptDict(obj)  # type: ignore
+
+
+def script(obj, optimize=None, _frames_up=0, _rcb=None, type_hint=None):
```
For empty dictionaries, I set the type to `Dict[str, Tensor]` if no type is passed in (consistent with the rest of TS). Why don't we just get rid of the this `type_hint` parameter?

SplitInfinity replied:
I removed it as discussed offline.

Commit:
Update on "[JIT] Add torch._C.ScriptDict"


**Summary**
This commit adds `torch._C.ScriptDict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch._C.ScriptDict` in TorchScript are visible in
Python even when it is not returned from the function. Instances can be
constructed by passing an instance of a Python dictionary to
`torch.jit.script`. In the case of an empty dictionary, its type is
assumed to be `Dict[str, Tensor]` to be consistent with the handling of
empty dictionaries in TorchScript source code.

`torch._C.ScriptDict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch._C.ScriptDict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

Differential Revision: [D27211605](https://our.internmc.facebook.com/intern/diff/D27211605)

[ghstack-poisoned]

gmagogsfm commented on torch/jit/_script.py:
```diff
@@ -21,6 +21,7 @@
 from torch.utils import set_module
 from torch.jit._recursive import ScriptMethodStub, wrap_cpp_module, infer_methods_to_compile
 from torch.nn import Module
+from torch.jit.annotations import ann_to_type
```
Minor lint error, please fix

Commit:
Update on "[JIT] Add torch._C.ScriptDict"


**Summary**
This commit adds `torch._C.ScriptDict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch._C.ScriptDict` in TorchScript are visible in
Python even when it is not returned from the function. Instances can be
constructed by passing an instance of a Python dictionary to
`torch.jit.script`. In the case of an empty dictionary, its type is
assumed to be `Dict[str, Tensor]` to be consistent with the handling of
empty dictionaries in TorchScript source code.

`torch.jit.dict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch._C.ScriptDict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

Differential Revision: [D27211605](https://our.internmc.facebook.com/intern/diff/D27211605)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add torch._C.ScriptDict"


**Summary**
This commit adds `torch._C.ScriptDict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch._C.ScriptDict` in TorchScript are visible in
Python even when it is not returned from the function. Instances can be
constructed by passing an instance of a Python dictionary to
`torch.jit.script`. In the case of an empty dictionary, its type is
assumed to be `Dict[str, Tensor]` to be consistent with the handling of
empty dictionaries in TorchScript source code.

`torch.jit.dict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch._C.ScriptDict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

Differential Revision: [D27211605](https://our.internmc.facebook.com/intern/diff/D27211605)

[ghstack-poisoned]

SplitInfinity commented:
@SplitInfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27211605).

Commit:
Update on "[JIT] Add torch._C.ScriptDict"


**Summary**
This commit adds `torch._C.ScriptDict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch._C.ScriptDict` in TorchScript are visible in
Python even when it is not returned from the function. Instances can be
constructed by passing an instance of a Python dictionary to
`torch.jit.script`. In the case of an empty dictionary, its type is
assumed to be `Dict[str, Tensor]` to be consistent with the handling of
empty dictionaries in TorchScript source code.

`torch.jit.dict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch._C.ScriptDict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

Differential Revision: [D27211605](https://our.internmc.facebook.com/intern/diff/D27211605)

[ghstack-poisoned]

SplitInfinity commented:
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27211605).

Commit:
Update on "[JIT] Add torch._C.ScriptDict"


**Summary**
This commit adds `torch._C.ScriptDict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch._C.ScriptDict` in TorchScript are visible in
Python even when it is not returned from the function. Instances can be
constructed by passing an instance of a Python dictionary to
`torch.jit.script`. In the case of an empty dictionary, its type is
assumed to be `Dict[str, Tensor]` to be consistent with the handling of
empty dictionaries in TorchScript source code.

`torch.jit.dict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch._C.ScriptDict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

Differential Revision: [D27211605](https://our.internmc.facebook.com/intern/diff/D27211605)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add torch._C.ScriptDict"


**Summary**
This commit adds `torch._C.ScriptDict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch._C.ScriptDict` in TorchScript are visible in
Python even when it is not returned from the function. Instances can be
constructed by passing an instance of a Python dictionary to
`torch.jit.script`. In the case of an empty dictionary, its type is
assumed to be `Dict[str, Tensor]` to be consistent with the handling of
empty dictionaries in TorchScript source code.

`torch.jit.dict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch._C.ScriptDict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

Differential Revision: [D27211605](https://our.internmc.facebook.com/intern/diff/D27211605)

[ghstack-poisoned]

SplitInfinity commented:
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27211605).

Commit:
Update on "[JIT] Add torch._C.ScriptDict"


**Summary**
This commit adds `torch._C.ScriptDict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch._C.ScriptDict` in TorchScript are visible in
Python even when it is not returned from the function. Instances can be
constructed by passing an instance of a Python dictionary to
`torch.jit.script`. In the case of an empty dictionary, its type is
assumed to be `Dict[str, Tensor]` to be consistent with the handling of
empty dictionaries in TorchScript source code.

`torch.jit.dict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch._C.ScriptDict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

Differential Revision: [D27211605](https://our.internmc.facebook.com/intern/diff/D27211605)

[ghstack-poisoned]

SplitInfinity commented:
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27211605).

Commit:
Update on "[JIT] Add torch._C.ScriptDict"


**Summary**
This commit adds `torch._C.ScriptDict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch._C.ScriptDict` in TorchScript are visible in
Python even when it is not returned from the function. Instances can be
constructed by passing an instance of a Python dictionary to
`torch.jit.script`. In the case of an empty dictionary, its type is
assumed to be `Dict[str, Tensor]` to be consistent with the handling of
empty dictionaries in TorchScript source code.

`torch.jit.dict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch._C.ScriptDict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

Differential Revision: [D27211605](https://our.internmc.facebook.com/intern/diff/D27211605)

[ghstack-poisoned]

SplitInfinity commented:
@SplitInfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27211605).
